src/mlmc/quantity_concept.py

```diff
@@ -1,115 +1,430 @@
+import abc
 import numpy as np
+import copy
+import operator
+from inspect import signature
+from memoization import cached
+from scipy import interpolate
+from typing import List, Tuple
+from mlmc.sample_storage import SampleStorage
+from mlmc.sim.simulation import QuantitySpec
+
+
+def _get_quantity_info(args_quantities, get_quantity_storage=False):
+    """
+    Get basic information about quantities - base QType, QuantityStorage
+    :param args_quantities: list of quantities and other pass argmuents, we expect at least one of the arguments is Quantity
+    :param get_quantity_storage: bool, if True return also quantity storage
+    :return: base QType, QuantityStorage
+    """
+    q_storage = None
+    # Either all quantities are BoolType or it is considered to be ScalarType
+    for quantity in args_quantities:
+        if isinstance(quantity, Quantity):
+            if get_quantity_storage:
+                q_storage = quantity.get_quantity_storage()
+            if type(quantity.qtype.base_qtype()) == ScalarType:
+                return ScalarType(), q_storage
+
+    return BoolType(), q_storage
+
+
+def _determine_qtype(quantities, method):
+    """
+    Determine QType from evaluation with given method and first few samples from storage
+    :param quantities: list of Quantities
+    :param method: ufunc function
+    :return: QType
+    """
+    chunks_quantity_level = [q.samples(level_id=0, i_chunk=0, n_samples=10) for q in quantities]
```

Comment:
This complicates the samples method, what about to introduce a separate method that gets just few samples for level 0. Would help if we can separate the recursion mechanism from the actual recursive call. We have to start with the simplification of the samples method.